zeebe/feel-tagged-parameters/src/main/java/io/camunda/zeebe/feel/tagged/impl/FromAiTaggedParameterExtractor.java

```diff
@@ -0,0 +1,166 @@
+
+      default ->
+          throw new IllegalArgumentException(
+              "Unsupported fromAi() function invocation: " + functionInvocation.params());
+    };
+  }
+
+  private TaggedParameter fromPositionalFunctionInvocationParams(final List<Exp> params) {
+    final Function<Integer, Exp> getParam =
+        index -> (params.size() > index ? params.get(index) : null);
+
+    return fromFunctionInvocationParams(
+        getParam.apply(0),
+        getParam.apply(1),
+        getParam.apply(2),
+        getParam.apply(3),
+        getParam.apply(4));
+  }
+
+  private TaggedParameter fromNamedFunctionInvocationParams(final Map<String, Exp> params) {
+    return fromFunctionInvocationParams(
+        params.get("value"),
+        params.get("description"),
+        params.get("type"),
+        params.get("schema"),
+        params.get("options"));
+  }
+
+  private TaggedParameter fromFunctionInvocationParams(
+      final Exp name, final Exp description, final Exp type, final Exp schema, final Exp options) {
+
+    final var parameterName = parameterName(name);
+    final var descriptionStr = asString(description, "description");
+    final var typeStr = asString(type, "type");
+    final var schemaMap = asMap(schema, "schema");
+    final var optionsMap = asMap(options, "options");
+
+    return new TaggedParameter(parameterName, descriptionStr, typeStr, schemaMap, optionsMap);
+  }
+
+  private String parameterName(final Exp value) {
+    if (!(value instanceof final Ref valueRef)) {
+      throw new IllegalArgumentException(
+          "Expected fromAi() parameter 'value' to be a reference (e.g. 'toolCall.customParameter'), but received %s."
+              .formatted(
+                  switch (value) {
+                    case final ConstString stringResult ->
+                        "string '%s'".formatted(stringResult.value());
+                    default -> typeMismatchExceptionValue(value);
+                  }));
+    }
+
+    return String.join(".", CollectionConverters.asJava(valueRef.names()));
+  }
+
+  private String asString(final Exp exp, final String parameterName) {
+    if (exp == null) {
+      return null;
+    }
+
+    if (!(exp instanceof final ConstString constString)) {
+      throw new IllegalArgumentException(
+          "Expected fromAi() parameter '%s' to be a string, but received '%s'."
+              .formatted(parameterName, typeMismatchExceptionValue(exp)));
+    }
+
+    return constString.value();
+  }
+
+  private Map<String, Object> asMap(final Exp exp, final String parameterName) {
+    if (exp == null) {
+      return null;
+    }
+
+    if (!(exp instanceof final ConstContext constContext)) {
+      throw new IllegalArgumentException(
+          "Expected fromAi() parameter '%s' to be a context (map), but received '%s'."
+              .formatted(parameterName, typeMismatchExceptionValue(exp)));
+    }
+
+    return convertContext(constContext);
```

Comment:
Previous versions (and the version in the connector) evaluated this expression to get to a final value. This made it possible to specify things like:

```
fromAi(toolCall.myValue, string join(["A", "simple", "value"], " "), "str" + "ing", context put({}, "enum", ["A", "B", "C"]), { optional: not(false) })
```

However, as the evaluation context is a different one than for normal FEEL expressions (so expected variables would not exist as this is done at deployment time) I removed that flexibility and constrained this to only support constant values. This makes this implementation a bit easier as we don't need to evaluate and we can just analyze the already parsed expression.